### ISSUE TITLE
handling changes in the structure of path.data when upgrading from 2.x to 5.x to 6.x

### ISSUE THREAD

**Issue by andyb-elastic** [elastic/elasticsearch#32661]
When a 2.x cluster is created, the structure of `path.data` has all contents for a node inside a directory named after the cluster name. This was changed in 5.x (https://github.com/elastic/elasticsearch/pull/18554) to remove the directory with the cluster name and move its contents up a level. A 5.x cluster will still read the 2.x structure correctly. In 6.x this backwards compatible behavior is removed (https://github.com/elastic/elasticsearch/pull/20433), and if a 6.x node is started with a data directory using the old 2.x structure, it will see it as if it was empty and ignore the existing data.

This change is mentioned [in the 5.0 breaking change docs](https://github.com/elastic/elasticsearch/blame/56318153c8cdac28c86e7d72f18194e2f7e0b7e4/docs/reference/migration/migrate_5_0/fs.asciidoc#L12-L25) and [in the 6.0 breaking change docs](https://github.com/elastic/elasticsearch/blame/6.x/docs/reference/migration/migrate_6_0/cluster.asciidoc#L4-L27). The 6.0 breaking change docs say

```
# Assuming path.data is /tmp/mydata
# No longer supported:

$ tree /tmp/mydata
/tmp/mydata
├── <cluster_name>
│   └── nodes
│       └── 0
│           └── <etc>

# Should be changed to:

$ tree /tmp/mydata
/tmp/mydata
├── nodes
│   └── 0
│       └── <etc>
```

It would be reasonable for a user to read this and not take away from it "you will need to manually move the `nodes` directory up a level inside `path.data` and remove the cluster name directory"

It doesn't look like the migration plugin for 2.x to 5.x checks for this. The upgrade assistant for 5.x to 6.x does warn you if you have your cluster name in `path.data`, but doesn't warn about your `path.data` contents using the 2.x and earlier structure. 

The outcome of this is that users upgrading from 2.x to 5.x to 6.x may reasonably not understand they have to manually make changes to the contents of `path.data`, and end up with a failed 5.x -> 6.x upgrade. It seems like users would benefit from

1. It being made explicit how they are expected to handle this change when upgrading. For example, the relevant breaking changes doc (should this be in the 5.0 or 6.0 docs?) could read something like "before upgrading you will need to move the `nodes` directory up one level so it is directly inside `path.data`, and remove the now-empty directory named after your cluster" and include some example commands

2. (This ship maybe already sailed) The 5.x to 6.x upgrade assistant warning if it finds that the contents of `path.data` are organized with the old 2.x structure - I also realize now that the upgrade assistant may not actually be able to do this since it's running in kibana

3. (Maybe excessive) 6.x and later nodes refusing to start if they find the old 2.x structure inside their `path.data`

**Comment by talevy** [comment]
> I also realize now that the upgrade assistant may not actually be able to do this since it's running in kibana

these warnings are actually done in Elasticsearch https://github.com/elastic/x-pack-elasticsearch/blob/5.6/plugin/src/main/java/org/elasticsearch/xpack/deprecation/NodeDeprecationChecks.java#L43

Do you think anything should be done beyond revamping the docs? I'm happy to add some extra wording here

**Comment by andyb-elastic** [comment]
Ideally it would be something to warn users about before upgrading but I'm not sure how helpful it'll be at this point - users who are on an already-released 5.6 version won't get the warning

So it seems like the right thing to do is just making the docs more explicit unless we want to have 6.x nodes check for it at startup. I labeled this discuss in case anyone has better ideas

**Comment by talevy** [comment]
@andyb-elastic isn't that deprecation check a warning that they will receive? Sounds like there may be a bug in the check? It should error out saying there is a cluster name in the data path

**Comment by andyb-elastic** [comment]
It appears in the 5.x deprecation log normally but I didn't see any warnings in the upgrade assistant - I'm not exactly sure what the relationship between the two is

**Comment by talevy** [comment]
did you launch the upgrade assistant with a mixed cluster? or when everything was 5.x?

**Comment by andyb-elastic** [comment]
This was a single-node 5.6.10 cluster

**Comment by bleskes** [comment]
we discussed this in fix it friday and we actually prefer option 3 - i.e., a 6.x node will refuse to start if the structure of the data folder is not what is expected (i.e. it has a nodes subfolder and that's it, potentially with some leniency for hidden folders for mac users). @talevy @andyb-elastic , please comment if that works for you.

**Comment by talevy** [comment]
@bleskes yeah, that sounds good

**Comment by andyb-elastic** [comment]
@bleskes I agree, that works. Thanks

**Comment by bleskes** [comment]
@talevy where do we stand on this one?

**Comment by talevy** [comment]
@bleskes regarding making it so that a 6.x node refuses to start up if it finds the old 2.x structure inside its `path.data`? Sounds like we concluded this is the action to take, but I have made no progress on this. I can get started on this if we would like it for 6.6.

**Comment by bleskes** [comment]
That would be great.

**Comment by talevy** [comment]
@bleskes cool. I'll add it to the forefront of my mind

**Comment by talevy** [comment]
closed by #36519 in ESv6.7
